src/utils/gadm_gpkg_processing.py

```diff
@@ -0,0 +1,200 @@
+            return True, dl_link, version 
+        else:
+            return True, None, version
+    else:
+        print('Did not recognise supplier.')
+        return False, None, None
+
+##
+#
+def download_url(url, save_path, chunk_size=128):
+    r = requests.get(url, stream=True)
+    with open(save_path, 'wb') as fd:
+        for chunk in r.iter_content(chunk_size=chunk_size):
+            fd.write(chunk)
+    #logger.info(f'Downloaded "{url}" to "{save_path}"')
+    print('Downloaded {0} to {1}'.format(url, save_path))
+
+
+##
+#
+def is_int(string):
+    try:
+        int(string)
+        return True
+    except ValueError:
+        return False
+
+##
+# Aggregate lower admin levels for specified country
+def get_country_admin_levels(gdf_world, country_aoi):
+    print('Checking {0} Admin Levels.'.format(country_aoi))
+    return_dict = {}
+    gdf = gdf_world.loc[gdf_world['GID_0'] == country_aoi].copy()
+    # Determine level of admin - check GID_* cols for None values.
+    cols = gdf.columns.values.tolist()
+    gdf_gid = gdf[fnmatch.filter(cols, 'GID_*')].copy()
+    gdf_gid.dropna(axis=1, how='all', inplace=True)
+    # Last list element, Last character
+    last_populated_gid = gdf_gid.columns.values.tolist()[-1][-1]
+    print('\tMax Admin Level {0}'.format(last_populated_gid))
+    # Drop all cols after first empty GID_* column.
+    drop_cols = []
+    for col in cols:
+        if is_int(col[-1]) is True and int(col[-1]) > int(last_populated_gid):
+            drop_cols.append(col)
+    gdf.drop(drop_cols, axis=1, inplace=True)
+    # and take out ID_* fields
+    gdf.drop(gdf[fnmatch.filter(gdf.columns.values.tolist(), 'ID_*')], 
+             axis=1, inplace=True)
+    # and take out REMARKS_*
+#    gdf.drop(gdf[fnmatch.filter(gdf.columns.values.tolist(), 'REMARKS_*')], 
+#             axis=1, inplace=True)
+    # and take out VALID*
+#    gdf.drop(gdf[fnmatch.filter(gdf.columns.values.tolist(), 'VALID*')], 
+#             axis=1, inplace=True)
+    # and the UID
+    gdf.drop(columns=['UID'], axis=1, inplace=True)
+    # anything else is associated with the last avialable admin level if
+    # populated.
+    gdf.dropna(axis=1, how='all', inplace=True)
+    #
+    level = 0
+    while level < int(last_populated_gid):
+        print('\tProcessing Level {0}'.format(level))
+        cols_agg = []
+        cols_required = []
+        for col in gdf.columns.values.tolist():
+            if is_int(col[-1]) is True and int(col[-1]) <= level:
+                cols_agg.append(col)
+        # Include the geometry column 
+        cols_required = list(cols_agg)
+        cols_required.append('geometry')
+        # Dissolve using GID_* only - assume all other fields to use First
+        # value. Was dropping features where looking to be specific and
+        # groupby more fields (that should all be the same)
+        cols_agg = [x for x in cols_agg if 'GID_' in x]
+        gdf_dslv = gdf[cols_required].dissolve(by=cols_agg)
```

Comment:
I noticed that there is a wolrd file available on GADM with the admin regions as separte layers. I think by using that file you could avoid having to reverse-engineer the higher level regions using `dissolve`. Is there an atvantage to using the single layer file?

Reply:
Good point, I missed that. I guess it boils down to whether there's a benefit from being in a position where we could probably extract everything that's required from GADM from a single geopackage / download vs using and managing multiple download links depending on the requirement.

Reply:
So the file I had in mind is actually a single geopackage: [this one here](https://biogeo.ucdavis.edu/data/gadm3.6/gadm36_levels_gpkg.zip), and each layer is a different admin level. With this file, you could use fiona to select only the admin 0 layer, and then work with that to create the surrounding area. 
I'm just concerned that using `dissolve` could be prone to error, e.g. what if a lower-level admin region is missing somewhere?

Reply:
In theory this would still work assuming the data is good as it would detect the lowest level available and work up from there.